### PR TITLE
Remove the FL_APP_BASE_BUCKET S3 dependency

### LINKED ISSUE #724: Remove the FL_APP_BASE_BUCKET S3 dependency
## Problem

The base FL application files in `fl-apps/` are currently published to S3 and consumed through `FL_APP_BASE_BUCKET`. This creates an avoidable AWS dependency for new FLIP deployments and requires a separate CI-driven publishing workflow whenever the application templates change.

The current design appears to be a legacy of the period when the base applications lived in separate repositories. They now live in the FLIP monorepo and can be versioned and deployed atomically with `flip-api`.

## Current behaviour

Only `flip-api` consumes the base bucket:

- The NVFlare and Flower bundlers list base application files in S3 and copy them into `FL_APP_DESTINATION_BUCKET/<model_id>`, alongside the user's scanned model files.
- Model creation downloads each backend's `required_files.json` from the base bucket.
- FL servers and clients do not access the base bucket. They receive presigned URLs for the completed bundle in the destination bucket and download those files over HTTPS.

The destination, scanned-model, results, and other runtime-data buckets remain necessary under the current AWS deployment. This issue concerns only repository-owned base application code.

## Why change it

- External users must provision and populate an undocumented S3 prefix before FLIP can bundle applications.
- Template changes require GitHub workflows, AWS OIDC permissions, bucket synchronisation, and environment-specific handling.
- The running `flip-api` image and S3 templates can drift independently.
- PR testing relies on publishing templates to a temporary S3 prefix and repointing configuration.
- Stale `FL_APP_BASE_BUCKET` configuration can break the required-files manifest pull.
- S3-to-S3 copying provides negligible value for the small base application payload; large user artifacts already remain in S3.

## Proposed approach

1. Include `fl-apps/` in the production and staging `flip-api` image.
2. Bind-mount `fl-apps/` into the development container for immediate local iteration.
3. Add an optional `FL_APP_BASE_DIR` setting that defaults to the baked-in path, allowing operators to mount alternative templates.
4. Replace base-bucket listing and S3-to-S3 copies with a local directory walk and uploads to `FL_APP_DESTINATION_BUCKET`.
5. Read `required_files.json` directly from the local base directory.
6. Keep the completed application bundle and all runtime/user data in the existing destination buckets.
7. Remove obsolete base-app S3 configuration, Terraform variables, sync workflows, PR override/cleanup workflows, and migration tooling references.

The Docker build context currently needs adjustment because `flip-api/` is used as the context while `fl-apps/` is at repository root. Options include widening the build context or using a named BuildKit additional context.

## Acceptance criteria

- [ ] NVFlare and Flower applications bundle successfully using local base files.
- [ ] User-submitted/scanned model files are still merged into the destination S3 bundle.
- [ ] `required_files.json` is loaded locally for both backends.
- [ ] Production and staging images contain the expected `fl-apps` revision.
- [ ] Development compose mounts local `fl-apps/` without an S3 sync step.
- [ ] A configurable local base directory supports operator-provided templates.
- [ ] `FL_APP_BASE_BUCKET` and `FL_APP_BASE_BUCKET_OVERRIDE` are removed after migration.
- [ ] Base-application S3 sync, PR publish, and cleanup workflows are removed.
- [ ] Relevant tests cover missing directories/files, nested paths, upload failures, and backend/job-type selection.
- [ ] Deployment and contributor documentation describe the local-template model.
- [ ] Existing deployments have a coordinated migration path so image rollout and configuration removal happen safely.

## Trade-off

Template-only production hotfixes will require rebuilding and deploying `flip-api` instead of syncing S3. This is preferable because the templates, manifest validation, and bundling logic are coupled and should be released as one versioned artifact.

### PR DESCRIPTION
**Pull request by atriaybagur**
Closes #724.

## Summary

The base FL application templates (`fl-apps/`) are no longer published to S3 and read through `FL_APP_BASE_BUCKET`. They are now **baked into the `flip-api` image** and read from a local directory (`FL_APP_BASE_DIR`, default `/app/fl-apps`, bind-mounted in dev). Only the *completed bundle* still goes to `FL_APP_DESTINATION_BUCKET`; all runtime/user data buckets are unchanged.

This removes an avoidable AWS dependency for new deployments and the separate CI publishing pipeline, and lets templates + manifest validation + bundling logic ship as one versioned artifact with `flip-api`.

## What changed

**flip-api**
- `bundle_nvflare_application` / `bundle_flower_application` walk `<FL_APP_BASE_DIR>/<backend>/<job_type>/` locally and **upload** the templates into `FL_APP_DESTINATION_BUCKET/<model_id>` (new `S3Client.upload_file`), instead of S3-to-S3 copies. User-submitted/scanned model files are still merged into each app. (Also drops the old redundant double base-copy and now validates required files *before* uploading, so a bad submission can't leave a partial bundle.)
- `required_files.json` is read directly from `<FL_APP_BASE_DIR>/<backend>/`; `pull_required_files.py` and the runtime S3-pull/assets machinery are removed.
- `FL_APP_BASE_BUCKET` removed from `Settings`; `FL_APP_BASE_DIR` added (optional, defaults to the baked-in path so operators can mount alternative templates).

**Build / deploy**
- `fl-apps/` is baked in via a **BuildKit named build context** (`fl_apps=../fl-apps`) since it sits outside `flip-api/`'s build context. Dev compose bind-mounts `../fl-apps` for live iteration; CI passes `--build-context fl_apps=../fl-apps` with `DOCKER_BUILDKIT=1`.
- Removed `FL_APP_BASE_BUCKET` from dev/prod compose and from the ECS task env + `fl_app_base_uri` local in Terraform.
- Deleted the 10 `fl-apps-push-s3-*` / `fl-apps-push-pr-s3-*` / `fl-apps-cleanup-pr-s3-*` workflows. The `fl-apps-check-required-files` and `fl-apps-check-tutorial-sync` validators stay.

**Docs**: `.env.development.example`, `flip-api/README.md`, `flip-utils/README.md`, `deploy/providers/AWS/README.md` (+ migration Makefile/verify script), `CONTRIBUTING.md`, `CLAUDE.md`/`AGENTS.md`, `fl-apps/check_required_files.sh`.

## Migration / rollout (existing deployments)

`Settings` uses `extra="ignore"`, so a stale `FL_APP_BASE_BUCKET` in an env file is harmless. Roll out **image first, then remove config**:
1. Build + deploy the new `flip-api` image (contains `fl-apps/`, reads locally).
2. Then apply the Terraform / config change that drops `FL_APP_BASE_BUCKET`.

The new image ignores `FL_APP_BASE_BUCKET` entirely, so the ordering is safe either way, but doing the image first avoids booting an old image after the mandatory setting is removed.

**Trade-off** (accepted on the issue): a template-only production hotfix now requires rebuilding + redeploying `flip-api` instead of an S3 sync — preferable because templates, manifest validation, and bundling are coupled and should release as one versioned artifact.

## Testing

- `flip-api` unit suite green (ruff + mypy + pytest, 1190 passed / 7 skipped). Bundler tests rewritten onto a local base-directory tree; new coverage for missing dir/file, nested paths, upload failure, and backend/job-type selection. `pull_required_files` test removed.
- Verified the BuildKit named-context mechanism (`COPY --from=fl_apps`) bakes the tree correctly (nvflare manifest + flower template present), and that dev compose renders the named context, the `/app/fl-apps` bind-mount, and no `FL_APP_BASE_BUCKET`.
- `required_files.json` generator runs with no drift.

## Acceptance Criteria

*Imported from issue #724*

- [x] NVFlare and Flower applications bundle successfully using local base files.
- [x] User-submitted/scanned model files are still merged into the destination S3 bundle.
- [x] `required_files.json` is loaded locally for both backends.
- [x] Production and staging images contain the expected `fl-apps` revision.
- [x] Development compose mounts local `fl-apps/` without an S3 sync step.
- [x] A configurable local base directory supports operator-provided templates.
- [x] `FL_APP_BASE_BUCKET` and `FL_APP_BASE_BUCKET_OVERRIDE` are removed after migration.
- [x] Base-application S3 sync, PR publish, and cleanup workflows are removed.
- [x] Relevant tests cover missing directories/files, nested paths, upload failures, and backend/job-type selection.
- [x] Deployment and contributor documentation describe the local-template model.
- [x] Existing deployments have a coordinated migration path so image rollout and configuration removal happen safely.

### CODE REVIEW

File: flip-api/src/flip_api/fl_services/services/fl_service.py [hunk 1]
```diff
@@ -42,6 +43,27 @@ class UnknownJobTypeError(Exception):
...
+        base_dir (Path): Root directory of a backend/job-type base application
+            (``<FL_APP_BASE_DIR>/<backend>/<job_type>``).
+
+    Returns:
+        list[str]: Sorted relative POSIX paths of every file under ``base_dir`` (nested paths
+        included). Empty if ``base_dir`` does not exist or contains no files.
+    """
+    if not base_dir.is_dir():
+        return []
+    return sorted(p.relative_to(base_dir).as_posix() for p in base_dir.rglob("*") if p.is_file())
```
[the remark this answers is not in the document]
Comment (atriaybagur) [reply flip-api/src/flip_api/fl_services/services/fl_service.py]: Addressed in 2dfc1f66: `list_local_base_files` now uses `os.walk(followlinks=False)` and skips symlinked files, so a stray/hostile symlink under an operator-provided `FL_APP_BASE_DIR` can no longer pull host files outside the template tree into the bundle. (The default baked-in `fl-apps/` tree has no symlinks; kept the fix cheap rather than adding a full realpath-containment check since the operator mounting the dir is already trusted.) Added a regression test.

File: flip-api/src/flip_api/utils/s3_client.py [hunk 2]
```diff
@@ -317,6 +318,33 @@ def object_exists(self, s3_path: str) -> bool:
...
+        try:
+            bucket, key = parse_s3_path(s3_path)
+            # boto3's managed upload wraps S3-side failures (AccessDenied, NoSuchBucket, …) in
+            # S3UploadFailedError — a boto3.exceptions type, not a botocore ClientError — so it must
+            # be caught explicitly alongside ClientError; OSError covers a local-file read failure.
+            self.client.upload_file(local_path, bucket, key)
+            logger.info(f"Successfully uploaded {local_path} to {s3_path}")
+        except (S3UploadFailedError, ClientError, OSError) as e:
+            logger.error(f"Error uploading {local_path} to {s3_path}: {e}")
+            raise Exception(f"Unable to upload file {local_path} to {s3_path}: {e}")
```
[the remark this answers is not in the document]
Comment (atriaybagur) [reply flip-api/src/flip_api/utils/s3_client.py]: Addressed in 2dfc1f66: added `from e` so the original error chains explicitly (matches `get_put_presigned_post`).

File: flip-api/src/flip_api/fl_services/services/fl_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (atriaybagur) [reply flip-api/src/flip_api/fl_services/services/fl_service.py]: Confirmed addressed and carried through the develop merge (2e00fa22): `list_local_base_files` skips symlinks via `os.walk(followlinks=False)` + a per-file `is_symlink()` guard, with a regression test. Resolving.

**Comment by atriaybagur** [comment]
### Update since opening — merged `develop` + validated e2e on both backends

**Merged `origin/develop`** (it had advanced 42 commits via the Ark+ PR #710). Conflict resolution + reconciliation:
- `deploy/providers/AWS/locals.tf`, `flip_api/.../fl_service.py`, and a modify/delete on a removed workflow — resolved in favour of the local-dir bundler.
- Re-applied the #724 removal onto base-bucket machinery that develop reintroduced: dropped `FL_APP_BASE_BUCKET_OVERRIDE` from `deploy/providers/AWS/{Makefile,variables.tf}` (the issue AC calls for removing it too).
- Reconciled `test_fl_service.py` where develop's new **server-checkpoint diversion** bundler tests overlapped my rewrite: converted the 3 checkpoint tests to the local-dir setup, and removed the 2 now-obsolete "delimiter-safe S3 prefix" tests (there's no sibling-prefix bleed with a local directory walk). The server-checkpoint feature itself merged cleanly and coexists with the local-dir bundler (`verify_bundle_paths` takes both `base_rel_paths` and `server_checkpoints`).

**Copilot review**: both comments (skip symlinks in the base walk; `from e` on the upload error) addressed in `2dfc1f66` and the threads resolved.

**End-to-end validation on a live dev stack — both backends PASS** (`RESULTS_UPLOADED`):
- **Flower** and **NVFLARE** each ran the full lifecycle (create project → cohort → DICOM pull → bundle → FL training → results download), reusing one project so DICOM pulled once.
- Confirmed the #724 path in flip-api logs: it walks the local `fl-apps` templates and `upload_file`s them into `…/app_destinations/<model_id>/` — e.g. NVFLARE logged `Uploading base /app/fl-apps/nvflare/standard/…` → `Bundle verification succeeded: 10 files present`. No S3 base-bucket listing anywhere.

CI is green (31 checks pass) and the branch is mergeable.

**Review by garciadias — APPROVED** [review]
Reviewed — clean removal of FL_APP_BASE_BUCKET S3 dependency. Templates + bundler + manifest now ship as one versioned flip-api artifact. Tests rewritten comprehensively onto local-tree model with good edge-case coverage (symlink skip, nested paths, upload failure propagation). Docs updated across compose, CI, Terraform, AGENTS.md, CLAUDE.md, READMEs, and tutorial troubleshooting guide. Migration path documented: image-first rollout, then config removal.

No blocking findings. LGTM.
